shared/team-building/service-tab-bar.desktop.tsx

```diff
@@ -4,106 +4,238 @@ import * as Styles from '../styles'
+      onChangeService: (service: ServiceIdWithContact) => void
+    }>
+  ) => (
+    <>
+      <Kb.Box2 direction="vertical" fullHeight={true} fullWidth={true} style={styles.moreNetworks0}>
+        <Kb.Box2
+          direction="vertical"
+          style={styles.moreNetworks1}
+          fullHeight={true}
+          centerChildren={true}
+          ref={props.setAttachmentRef}
+        >
+          <Kb.WithTooltip text="More networks" containerStyle={styles.moreNetworks2}>
+            <Kb.ClickableBox onClick={props.toggleShowingMenu} style={styles.moreNetworks3}>
+              <Kb.Text type="BodyBigExtrabold" style={styles.moreText}>
+                •••
+              </Kb.Text>
+            </Kb.ClickableBox>
+          </Kb.WithTooltip>
+        </Kb.Box2>
+        <Kb.Box2 direction="horizontal" fullWidth={true} style={styles.inactiveTabBar} />
+      </Kb.Box2>
+      <Kb.FloatingMenu
+        attachTo={props.getAttachmentRef}
+        closeOnSelect={true}
+        items={props.services.map(service => ({
+          onClick: () => props.onChangeService(service),
+          title: service,
+          view: <MoreNetworkItem service={service} />,
+        }))}
+        onHidden={props.toggleShowingMenu}
+        visible={props.showingMenu}
       />
-    ))}
+    </>
+  )
+)
+
+const MoreNetworkItem = (props: {service: ServiceIdWithContact}) => (
+  <Kb.Box2 direction="horizontal" fullHeight={true} alignItems="center">
+    <Kb.Icon style={styles.moreNetworkItemIcon} fontSize={16} type={serviceIdToIconFont(props.service)} />
+    <Kb.Text type="Body">{serviceIdToLongLabel(props.service)}</Kb.Text>
   </Kb.Box2>
 )
 
-const styles = Styles.styleSheetCreate({
+const undefToNull = (n: number | undefined | null): number | null => (n === undefined ? null : n)
+
+export const ServiceTabBar = (props: Props) => {
+  const [
+    lastSelectedUnlockedService,
+    setLastSelectedUnlockedDevice,
+  ] = React.useState<ServiceIdWithContact | null>(null)
+  const nLocked = 3 // Services always out front on the left. Add one to get the number out front.
+  const lockedServices = Constants.services.slice(0, nLocked)
+  let frontServices = lockedServices
+  let lastSelectedUnlockedServiceChanged = false
+  if (Constants.services.indexOf(props.selectedService) < nLocked) {
+    // Selected service is locked
+    if (lastSelectedUnlockedService === null) {
+      frontServices = Constants.services.slice(0, nLocked + 1)
+    } else {
+      frontServices = lockedServices.concat([lastSelectedUnlockedService])
+    }
+  } else {
+    frontServices = lockedServices.concat([props.selectedService])
+    if (lastSelectedUnlockedService !== props.selectedService) {
+      lastSelectedUnlockedServiceChanged = true
+    }
+  }
+  React.useEffect(() => {
+    if (lastSelectedUnlockedServiceChanged) {
+      setLastSelectedUnlockedDevice(props.selectedService)
+    }
+  }, [lastSelectedUnlockedServiceChanged, props.selectedService])
```

Comment:
Here's a shot with `useEffect`. I just had to see if it would work. If you still prefer putting the setState in onClick lmk

Reply:
I think I'd prefer setting the state in onClick. I believe the difference is:

### useEffect
1. User selects icon
2. `onChangeService` fires, dispatches action
3. Redux state updated
4. New selected service passed down from store to `ServiceTabBar`, triggering re-render
5. `lastSelectedUnlockedServiceChanged` set to true
6. `setLastSelectedUnlockedService` queued and triggered after render completes, triggering a re-render
7.  `lastSelectedUnlockedServiceChanged` set to false, no re-render

### onClick
1. User selects icon
2. `setLastSelectedUnlockedService` queued
2. `onChangeService` fires, dispatches action
3. Redux state updated
4. New selected service passed down from store to `ServiceTabBar`
5. State of `ServiceTabBar` updated
6. I believe 4 and 5 will be batched, triggering only one re-render?


I could totally be wrong though. I suppose an alternative to either of these approaches to would be to just put the last selected unlocked service in redux and just treat it as another prop here. This has the advantage of persisting state through loads of the new group. It should also trigger only one re-render.

Reply:
This line all by itself causes render to happen twice every time a service is changed.
```React.useEffect(() => { }, [props.selectedService])```

Reply:
You're right that calling setState in useEffect world caused it to render 4x instead of 2x times when clicking something that changed the state.

Reply:
This line also causes render to happen twice all by itself. Must be something about hooks
```const [state, setState] = React.useState(null)```